firewood/src/merkle.rs

```diff
@@ -864,83 +864,49 @@ impl<S: ShaleStore<Node> + Send + Sync> Merkle<S> {
         key: K,
         root: DiskAddress,
     ) -> Result<Option<Vec<u8>>, MerkleError> {
-        let mut chunks = vec![0];
-        chunks.extend(key.as_ref().iter().copied().flat_map(to_nibble_array));
-
         if root.is_null() {
             return Ok(None);
         }
 
-        let mut deleted = Vec::new();
-        let mut parents: Vec<(ObjRef<Node>, _)> = Vec::new();
-        let mut u_ref = self.get_node(root)?;
-        let mut nskip = 0;
-        let mut found = None;
+        let (found, parents, deleted) = {
+            let mut parents = Vec::new();
+            let node_ref = self.get_node_by_key(self.get_node(root)?, key, |node_ref, nib| {
```

Comment:
I don't disagree with your reasoning, but I can't see a clean way to implement both methods without duplicating the logic. 

It's literally one line in the middle of loop, but the issue is that the one line is a mutation that depends on external (to the loop) state. 

Unless you're cool with `get_node_by_key_with_callback` as the actual implementation, then I could wrap that in `get_node_by_key` and `get_node_and_parents_by_key`. But in order to follow those methods, you're just going to look at the callback method anyway. It's more cognitive overhead to have three methods compared to one. If these methods were pub, it would be a different story. 

Unless you can think of a way to implement both methods without duplicating code and without using a closure, I think the change should stay.

Happy to discuss further though

Reply:
I don't think generic-bloat is relevant in private methods. There's almost no way the compiler doesn't just inline private methods like this.

Reply:
Maybe add a parameter `Option<&mut parents>` and pass `None` when you don't want them

Reply:
addressed:
https://github.com/ava-labs/firewood/pull/329/commits/6952f067e2ae25e0de4a42cf184f92aafff2b84a

Reply:
🚀 so much better!!!

Reply:
Agreed!